### PR TITLE
json.nim: avoid recursions that can lead to stack overflows [backport:1.0]

### PR DESCRIPTION
**Pull request by Araq**

### CODE REVIEW

File: lib/pure/json.nim [hunk 1]
```diff
@@ -884,28 +884,63 @@ proc parseJson(p: var JsonParser; rawIntegers, rawFloats: bool): JsonNode =
...
+          raiseParseErr(p, "{")
+        if p.tok == tkComma:
+          discard getTok(p)
+      of tkBracketRi:
+        if insertPos.len > 0 and insertPos[^1].kind == JArray:
+          result = insertPos.pop
+          discard getTok(p)
+          if insertPos.len == 0: break
+        else:
+          raiseParseErr(p, "{")
```
Comment (planetis-m) [inline lib/pure/json.nim:939]:
```suggestion
          raiseParseErr(p, "[")
```

File: lib/pure/json.nim [hunk 2]
```diff
@@ -884,28 +884,63 @@ proc parseJson(p: var JsonParser; rawIntegers, rawFloats: bool): JsonNode =
...
+        if p.tok == tkComma:
+          discard getTok(p)
+      else:
+        raiseParseErr(p, "{")
```
Comment (planetis-m) [inline lib/pure/json.nim:943]:
```suggestion
        raiseParseErr(p, "{ or [")
```

**Review by planetis-m — COMMENTED** [review]
Works as expected, now I get your seq PatchPos idea https://github.com/planetis-m/packedjson2/commit/761e75bfbc1730bf2ab73c502912ce7151667491 :)

**Comment by planetis-m** [comment]
Actually errors are completely skewed (even before the patch `{"a":}` outputs: `{` expected), scratch my previous comments they don't improve anything.

**Comment by Araq** [comment]
Too risky and too ugly. Instead we'll introduce a recursion limit.

**Comment by Araq** [comment]
Note: I don't want to delete the branch yet.

**Pull request closed without merging**